association_measures/measures.py

```diff
@@ -5,149 +5,202 @@
+        return ntok // ktok
+    else:
+        return 0
+
+
+def phi(o, e):
+    """
+    Calculate phi(o,e):=o*log(o/e) with lim_{o↓0} phi(o,e)=0
+
+    :param int o: observed frequency
+    :param float e: expected frequency
+    :return: phi
+    :rtype: float
+    """
+
+    if o == 0:
+        return 0
+    elif e == 0:
+        raise ValueError("observed value must be 0 if expected value is zero")
+    else:
+        return o * np.log(o / e)
 
 
-def z_score(df):
+def z_score(freq):
     """
     Calculate z-score
 
-    :param pandas.DataFrame df: Pandas Dataframe containing O11 and E11
-    :return: pandas.Series containing the Z-Score for each token
-    :rtype: pandas.Series
+    :param dict freq: dictionary-like object with O11 and E11
+    :return: z-score
+    :rtype: float
     """
 
-    if 'E11' not in df.columns:
-        return np.nan
-
-    res = (df['O11'] - df['E11']) / np.sqrt(df['E11'])
+    am = (freq['O11'] - freq['E11']) / np.sqrt(freq['E11'])
 
-    return pd.Series(data=res)
+    return am
 
 
-def t_score(df):
+def t_score(freq):
     """
     Calculate t-score
 
-    :param pandas.DataFrame df: Pandas Dataframe containing O11 and E11
-    :return: pandas.Series containing the T-Score for each token
-    :rtype: pandas.Series
+    :param dict freq: dictionary-like object with O11 and E11
+    :return: t-score
+    :rtype: float
     """
 
-    if 'E11' not in df.columns:
-        return np.nan
-
-    res = (df['O11'] - df['E11']) / np.sqrt(df['O11'])
+    am = (freq['O11'] - freq['E11']) / np.sqrt(freq['O11'])
 
-    return pd.Series(data=res)
+    return am
 
 
-def mutual_information(df):
+def mutual_information(freq):
     """
     Calculate Mutual Information
 
-    :param pandas.DataFrame df: Pandas Dataframe containing O11 and E11
-    :return: pandas.Series containing the Mutual Information score for
-    each token
-    :rtype: pandas.Series
+    :param dict freq: dictionary-like object with O11 and E11
+    :return: mutual_information
+    :rtype: float
     """
 
-    if 'E11' not in df.columns:
-        return np.nan
-
-    diff = df['E11'].replace(0, np.nan) / df['O11'].replace(0, np.nan)
-    res = np.log(diff.replace(0.0, np.nan))
+    am = np.log10(freq['O11'] / freq['E11'])
 
-    return pd.Series(data=res)
+    return am
 
 
-def dice(df):
+def dice(freq):
     """
     Calculate Dice coefficient
 
-    :param pandas.DataFrame df: Pandas Dataframe containing O11, f1 and f2
-    :return: pandas.Series containing the Dice coefficient for each token
-    :rtype: pandas.Series
+    :param dict freq: dictionary-like object with O11, O12, O21
+    :return: dice
+    :rtype: float
     """
 
-    if 'O11' not in df.columns:
-        return np.nan
-
-    res = (2 * df['O11']) / (df['f1'] + df['f2'])
+    am = (2 * freq['O11']) / (2 * freq['O11'] + freq['O12'] + freq['O21'])
 
-    return pd.Series(data=res)
+    return am
 
 
-def log_likelihood(df):
+def log_likelihood(freq, signed=True):
```

Comment:
I would like these functions to be as polymorphic as possible, can we maybe not have a parameter but just return an extra column with the signed valued?

Reply:
I don't really understand the issue of polymorphism here, but I don't really mind.

I just removed the unsigned version, nobody should use this one anyway …